jmespath-core/src/main/java/io/burt/jmespath/Adapter.java

```diff
@@ -82,6 +94,12 @@
   T getProperty(T value, String name);
 
   /**
+   * Returns all of the property names of the given object, or an empty list
+   * when the given value is not an object.
+   */
+  Collection<String> getPropertyNames(T value);
```

Comment:
Would it not make more sense to return `T` here, which should then be a "JSON array of strings". Or at the very least `List<T>` for a list of "JSON strings". It seems this function would only be used in a context where the first thing that will happen is that the results are converted to JSON objects (and typically the implementation of it would have to convert JSON strings into Java strings).

I see that getProperty above uses `String` for name, which might make it counter-intuitive for this to not return `String`. I do wonder if it wouldn't be better to use "JSON strings" in both cases, though.